File: addTwoNumbers.py
from typing import Optional

from pylist import ListNode


def reverse(x):
    if x == None or x.next == None:
        return x
    p = x
    q = x.next
    p.next = None

    while q != None:
        r = q.next
        q.next = p
        p = q
        q = r

    return p


# addTwoNumbers
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        l1 = reverse(l1)
        l2 = reverse(l2)
        p = l1
        q = l2
        out = None
        start = None
        j = 0

        while p != None and q != None:
            now = p.val + q.val + j
            j = now // 10
            now = now % 10
            tmp = ListNode(now)

            if out != None:
                out.next = tmp

            if out == None:
                start = tmp
            out = tmp
            p = p.next
            q = q.next

        while p != None:
            now = p.val + j
            j = now // 10
            now = now % 10
            tmp = ListNode(now)

            if out != None:
                out.next = tmp

            if out == None:
                start = tmp
            out = tmp
            p = p.next

        while q != None:
            now = q.val + j
            j = now // 10
            now = now % 10
            tmp = ListNode(now)

            if out != None:
                out.next = tmp

            if out == None:
                start = tmp
            out = tmp
            q = q.next

        if j != 0:
            tmp = ListNode(j)
            out.next = tmp
            out = tmp

        return reverse(start)


l1 = [7, 2, 4, 3]
l2 = [5, 6, 4]
l1 = [0]
l2 = [0]
l1 = ListNode.fromList(l1)
l2 = ListNode.fromList(l2)
sl = Solution()
out = sl.addTwoNumbers(l1, l2)
ListNode.travel(out)


# 2. 两数相加
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode],
                      l2: Optional[ListNode]) -> Optional[ListNode]:
        p = l1
        q = l2
        out = None
        start = None
        j = 0

        while p != None and q != None:
            now = p.val + q.val + j
            j = now // 10
            now = now % 10
            tmp = ListNode(now)

            if out != None:
                out.next = tmp

            if out == None:
                start = tmp
            out = tmp
            p = p.next
            q = q.next

        while p != None:
            now = p.val + j
            j = now // 10
            now = now % 10
            tmp = ListNode(now)

            if out != None:
                out.next = tmp

            if out == None:
                start = tmp
            out = tmp
            p = p.next

        while q != None:
            now = q.val + j
            j = now // 10
            now = now % 10
            tmp = ListNode(now)

            if out != None:
                out.next = tmp

            if out == None:
                start = tmp
            out = tmp
            q = q.next

        if j != 0:
            tmp = ListNode(j)
            out.next = tmp
            out = tmp

        return start


l1 = [2, 4, 3]
l2 = [5, 6, 4]
l1 = ListNode.fromList(l1)
l2 = ListNode.fromList(l2)
ListNode.travel(Solution().addTwoNumbers(l1, l2))

File: addTwoNumbers.c
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */

#include"ctools.h"
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include"clist.h"

//#define testMod
#ifdef testMod
void test() 
{

}
#endif

#define max(a,b) (((a)>(b))?(a):(b))
#define min(a,b) (((a)<(b))?(a):(b))

#ifndef testMod
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    int carry=0;
    int now;
    struct ListNode* out=NULL;
    struct ListNode* tail=NULL;
    while(l1!=NULL||l2!=NULL||carry!=0)
    {
        now=carry;
        if(l1!=NULL)
        {
            now+=l1->val;
            l1=l1->next;
        }
        if(l2!=NULL)
        {
            now+=l2->val;
            l2=l2->next;
        }
        carry=now/10;
        now=now%10;
        if(tail==NULL)
        {
            out=getListNode(now);
            tail=out;
        }
        else
        {
            tail->next=getListNode(now);
            tail=tail->next;
        }
    }
    return out;
}
#endif

int main()
{
#ifdef testMod
    test();
#endif
#ifndef testMod
    struct ListNode* head1=getListNode(2);
    head1->next=getListNode(4);
    head1->next->next=getListNode(3);
    struct ListNode* head2=getListNode(5);
    head2->next=getListNode(6);
    head2->next->next=getListNode(7);
    struct ListNode* out=addTwoNumbers(head1,head2);
    travelList(out);
#endif
    return 0;
}
    
